### ISSUE TITLE
Cannot Access Costumed Metrics in CPA

### ISSUE THREAD

**Issue by Zongshun96** [jthomperoo/custom-pod-autoscaler#138]
I was studying to use costumed metrics in this CPA implementation but having trouble to access the metric in CPA.

This is the error I am receiving in the log of CPA pod, while I can see the metric in prometheus dashboard. Any suggestion?
```
I0523 17:09:20.728508       1 main.go:144] Custom Pod Autoscaler version: v2.0.0
I0523 17:09:20.730899       1 main.go:236] Waiting 1 milliseconds before starting autoscaler
I0523 17:09:20.731062       1 main.go:294] Starting API using HTTP
I0523 17:09:20.732413       1 main.go:245] Starting autoscaler
E0523 17:09:30.737986       1 main.go:277] invalid metrics (1 invalid out of 1), first error is: failed to get pods metric: unable to get metric demo_app_button_clicks_per_second: unable to fetch metrics from custom metrics API: pods.custom.metrics.k8s.io "*" is forbidden: User "system:serviceaccount:default:prometheus-cpa-demo" cannot get resource "pods/demo_app_button_clicks_per_second" in API group "custom.metrics.k8s.io" in the namespace "default"
```

cpa.yaml
```yaml
apiVersion: custompodautoscaler.com/v1
kind: CustomPodAutoscaler
metadata:
  name: prometheus-cpa-demo
spec:
  template:
    spec:
      containers:
      - name: prometheus-cpa-demo
        image: prometheus-cpa-demo:latest
        imagePullPolicy: IfNotPresent
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: prometheus-demo-app
  roleRequiresMetricsServer: true
  config:
    - name: interval
      value: "10000"
```

config.yaml
```yaml
evaluate:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/evaluate.py"
metric:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/metric.py"
kubernetesMetricSpecs:
  - type: Pods
    pods:
      metric:
        name: demo_app_button_clicks_per_second
      target:
        type: AverageValue
        averageValue: 10m
requireKubernetesMetrics: true
runMode: "per-resource"
```

deployment.yaml
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: prometheus-demo-app
  labels:
    app: prometheus-demo-app
spec:
  selector:
    matchLabels:
      app: prometheus-demo-app
  template:
    metadata:
      labels:
        app: prometheus-demo-app
    spec:
      containers:
      - name: prometheus-demo-app
        image: flipstone42/k8s-prometheus-custom-scaling:latest
        resources:
          limits:
            memory: "128Mi"
            cpu: "100m"
        ports:
        - containerPort: 8000
          name: http
```

Inside the deployment container. 
The python code is from the author of the tutorial mentioned below.
```python
from flask import Flask, render_templatefrom prometheus_flask_exporter.multiprocess import GunicornInternalPrometheusMetrics
application = Flask(__name__, static_folder="./public", template_folder="./templates")metrics = GunicornInternalPrometheusMetrics(application)

@application.route("/")def index():    return render_template("index.html")

@application.route("/click-button", methods=["POST"])@metrics.counter("demo_app_button_clicks", "Number of button presses by user")def web_button():    return {}
```

I was following on this tutorial which was originally for HPA and trying to adapt to CPA.
`https://levelup.gitconnected.com/building-kubernetes-apps-with-custom-scaling-a-gentle-introduction-a332d7ebc795`

Thanks,

**Comment by jthomperoo** [comment]
Hey, this looks like it might be a bug with the [Custom Pod Autoscaler Operator](https://github.com/jthomperoo/custom-pod-autoscaler-operator) - from a quick first look it seems that the role being generated for the CPA doesn't have the right permissions for the metrics-server. I'm going to investigate a bit more and if this is right I'll make a new issue under that repo to track it - there should be an easy enough workaround too if this is the problem so I'll add it onto this ticket to try and get around this blocking you.
Thanks!

**Comment by jthomperoo** [comment]
As a workaround to see if this fixes it for you, could you try and modify your YAML, update your `cpa.yaml` to be like this:

```yaml
apiVersion: custompodautoscaler.com/v1
kind: CustomPodAutoscaler
metadata:
  name: prometheus-cpa-demo
spec:
  template:
    spec:
      containers:
      - name: prometheus-cpa-demo
        image: prometheus-cpa-demo:latest
        imagePullPolicy: IfNotPresent
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: prometheus-demo-app
  provisionRole: false
  config:
    - name: interval
      value: "10000"
```

And then add a new `role.yaml` to deploy:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: prometheus-cpa-demo
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - replicationcontrollers
  - replicationcontrollers/scale
  verbs:
  - '*'
- apiGroups:
  - apps
  resources:
  - deployments
  - deployments/scale
  - replicasets
  - replicasets/scale
  - statefulsets
  - statefulsets/scale
  verbs:
  - '*'
- apiGroups:
  - metrics.k8s.io
  resources:
  - '*'
  verbs:
  - '*'
- apiGroups:
  - custom.metrics.k8s.io
  resources:
  - '*'
  verbs:
  - '*'
- apiGroups:
  - external.metrics.k8s.io
  resources:
  - '*'
  verbs:
  - '*'
```

This workaround uses the `provisionRole: false` flag to let us provide our own role, which will have the permissions we need to access the custom metrics API - let me know if this helps, I'll try and recreate and test myself too.

**Comment by Zongshun96** [comment]
It works! Thank you!
Is there any good tutorial about Roles in k8s and how to configure it?

Thanks,

**Comment by jthomperoo** [comment]
Great, glad to hear that!

With this it's actually a bug you've found in the Custom Pod Autoscaler Operator, so I've created a new issue to fix this: https://github.com/jthomperoo/custom-pod-autoscaler-operator/issues/79

The workaround makes use of the custom resource configurations which you can find documented here:
https://github.com/jthomperoo/custom-pod-autoscaler-operator/blob/master/USAGE.md#using-custom-resources

Once this is merged in and a new version is cut (probably `v1.1.1`) you wont need this workaround anymore and your original configuration should work.

Thank you for finding this, it really helps to improve this project, closing this issue to track under the operator repo.
